### ISSUE TITLE
kotlin android: CodeQL detected code written in Java/Kotlin but could not process any of it

### ISSUE THREAD

**Issue by Goudan932** [github/codeql#17705]
I got message:**CodeQL detected code written in Java/Kotlin but could not process any of it**, when I run commend:
`codeql database create database --language=java-kotlin --overwrite`
I have already reviewed the historical issue records，but can not reslove this problem.
The log in terminal as:

> owenrwang@OWENRWANG-MC2 testcodeql2 % codeql database create database --language=java-kotlin --overwrite
Initializing database at /Users/owenrwang/AndroidStudioProjects/testcodeql2/database.
Running build command: [/Users/owenrwang/codeql/java/tools/autobuild.sh]
Running command in /Users/owenrwang/AndroidStudioProjects/testcodeql2: [/Users/owenrwang/codeql/java/tools/autobuild.sh]
[2024-10-09 11:27:44] [build-stdout] [2024-10-09 11:27:44] Build directory is /Users/owenrwang/AndroidStudioProjects/testcodeql2/.
[2024-10-09 11:27:44] [build-stdout] [2024-10-09 11:27:44] Found no Java toolchains in a toolchains file, inspecting environment variables...
[2024-10-09 11:27:44] [build-stdout] [2024-10-09 11:27:44] Found no Java toolchains in a toolchains file, inspecting environment variables...
[2024-10-09 11:27:45] [build-stdout] [2024-10-09 11:27:45] The Android plugin version in use recommends Java version 17.0.0.
[2024-10-09 11:27:45] [build-stdout] [2024-10-09 11:27:45] [WARN] The suggested lower bound Java version is 17.0.0, but no version that is equal to or greater than that is available on this system.
[2024-10-09 11:27:45] [build-stdout] [2024-10-09 11:27:45] Identified 20.0.0 as the newest version of Java supported by the version of Gradle in use. The system default of 17.0.12 is compatible with this.
[2024-10-09 11:27:45] [build-stdout] [2024-10-09 11:27:45] [autobuild] > ./gradlew --no-daemon -S -Dorg.gradle.dependency.verification=off -Dorg.gradle.warning.mode=none --init-script "/var/folders/98/0y5b_v_57lng7z5_9_2zxwg00000gn/T/semmleTempDir-26/init.gradle" clean
[2024-10-09 11:27:48] [build-stdout] [2024-10-09 11:27:48] [autobuild] To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.4/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
[2024-10-09 11:27:49] [build-stdout] [2024-10-09 11:27:49] [autobuild] Daemon will be stopped at the end of the build 
[2024-10-09 11:27:52] [build-stdout] [2024-10-09 11:27:52] [autobuild] > Task :app:clean UP-TO-DATE
[2024-10-09 11:27:52] [build-stdout] [2024-10-09 11:27:52] [autobuild] BUILD SUCCESSFUL in 3s
[2024-10-09 11:27:52] [build-stdout] [2024-10-09 11:27:52] [autobuild] 1 actionable task: 1 up-to-date
[2024-10-09 11:27:52] [build-stdout] [2024-10-09 11:27:52] [autobuild] > ./gradlew --no-daemon -S -Dorg.gradle.dependency.verification=off -Dorg.gradle.warning.mode=none --init-script "/var/folders/98/0y5b_v_57lng7z5_9_2zxwg00000gn/T/semmleTempDir-26/init.gradle" testClasses
[2024-10-09 11:27:53] [build-stdout] [2024-10-09 11:27:53] [autobuild] To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.4/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
[2024-10-09 11:27:53] [build-stdout] [2024-10-09 11:27:53] [autobuild] Daemon will be stopped at the end of the build 
[2024-10-09 11:27:56] [build-stdout] [2024-10-09 11:27:56] [autobuild] > Task :testClasses UP-TO-DATE
[2024-10-09 11:27:56] [build-stdout] [2024-10-09 11:27:56] [autobuild] BUILD SUCCESSFUL in 3s
Finalizing database at /Users/owenrwang/AndroidStudioProjects/testcodeql2/database.
CodeQL detected code written in Java/Kotlin but could not process any of it. For more information, review our troubleshooting guide at https://gh.io/troubleshooting-code-scanning/no-source-code-seen-during-build.

The files 'build-tracer.log' and 'database-create log' have been uploaded as attachments. Please refer to them.
[build-tracer.log](https://github.com/user-attachments/files/17301578/build-tracer.log)
[database-create-20241009.112742.502.log](https://github.com/user-attachments/files/17301581/database-create-20241009.112742.502.log)

The project I am using is the simplest Android project newly created with Android Studio. It has already been compiled with Android Studio and successfully installed on the phone.I only added two functions in MainActivity.kt for testing the ‘codeql database analyze’ commands, and to ensure the 'codeql database create' command executes properly, I added the following simple configuration in build.gradle: 
`task testClasses() {
}`
 (Without this configuration, an error occurs when executing 'codeql database create': [ERROR] Spawned process exited abnormally (code 1; tried to run: [./gradlew, --no-daemon, -S, -Dorg.gradle.dependency.verification=off, -Dorg.gradle.warning.mode=none, --init-script, /var/folders/98/0y5b_v_57lng7z5_9_2zxwg00000gn/T/semmleTempDir-27/init.gradle, testClasses])).

I have already tried '--build-mode none' when create database, but There is no my source code in '/Users/owenrwang/AndroidStudioProjects/testcodeql2/database/src.zip', and can not analyze my code when I run 'codeql database analyze' command. So I think maybe I need to build the code when executing the create database command in order for CodeQL to recognize the code I wrote, right?

Additional information：
**codeql --version**  result:
`
CodeQL command-line toolchain release 2.19.0.
Copyright (C) 2019-2024 GitHub, Inc.
Unpacked in: /Users/owenrwang/codeql
   Analysis results depend critically on separately distributed query and
   extractor modules. To list modules that are visible to the toolchain,
   use 'codeql resolve qlpacks' and 'codeql resolve languages'.`

**java --version** result:
`java 17.0.12 2024-07-16 LTS
Java(TM) SE Runtime Environment (build 17.0.12+8-LTS-286)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.12+8-LTS-286, mixed mode, sharing)`

**Comment by redsun82** [comment]
👋 @Goudan932 thanks for reaching out to us!

From what you wrote it seems to me:
* before you added `task testClasses() { }` the java/kotlin autobuild could not automatically find what to build, which was causing the error you mention
* after you added it, autobuild did pick that configuration up, but as that configuration is empty, nothing was actually built with led to no code being extracted in the database, leading to the "CodeQL detected code written in Java/Kotlin but could not process any of it" error

I will communicate with the appropriate team, as autobuild should probably be able to pick up an almost vanilla Android project. In the meantime, you should be able to proceed if you find the actual gradle invocation used to do the build. In that case `codeql database create database --language=java-kotlin --overwrite --command "gradle ???"` should unlock you.



> I have already tried '--build-mode none' when create database, but There is no my source code in '/Users/owenrwang/AndroidStudioProjects/testcodeql2/database/src.zip', and can not analyze my code when I run 'codeql database analyze' command. So I think maybe I need to build the code when executing the create database command in order for CodeQL to recognize the code I wrote, right?

Indeed, `--build-mode none` is not supported (yet!) for kotlin analysis and you do need to build the code.

**Comment by Goudan932** [comment]
@redsun82  Thank you very much for your response.
I have tried using the --command option to create the database:
`codeql database create database --language=java-kotlin --overwrite --command="./gradlew assembleDebug"`
However, it still prompts "CodeQL detected code written in Java/Kotlin but could not process any of it."

After executing the above command, the corresponding Kotlin compiled classes and .apk files were indeed generated under the project. 
Below is the generated dex file:
<img width="1519" alt="generatedDex" src="https://github.com/user-attachments/assets/c4226547-3a9a-415c-b4e6-ec5c940caddb">

Below is the class file generated from the Kotlin code: 
<img width="1581" alt="generatedClasses" src="https://github.com/user-attachments/assets/27f1996b-fb83-4908-9603-fc8a34154508">

Although everything compiled correctly, it seems that CodeQL did not create database by these artifacts (or the analysis failed?).

To troubleshoot whether it is an issue with Kotlin, I created a very simple pure Java Android project, but it still prompts "CodeQL detected code written in Java/Kotlin but could not process any of it."

The command is also as follows:
`codeql database create database --language=java-kotlin --overwrite --command="./gradlew assembleDebug"`
The terminal output is as follows:

> Initializing database at /Users/owenrwang/testqlwithjava/database.
Running build command: [./gradlew, assembleDebug]
Running command in /Users/owenrwang/testqlwithjava: [./gradlew, assembleDebug]
[2024-10-09 17:06:44] [build-stdout] > Task :app:preBuild UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:preDebugBuild UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugNativeDebugMetadata NO-SOURCE
[2024-10-09 17:06:44] [build-stdout] > Task :app:javaPreCompileDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:checkDebugAarMetadata UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:generateDebugResValues UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mapDebugSourceSetPaths UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:generateDebugResources UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugResources UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:packageDebugResources UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:parseDebugLocalResources UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:createDebugCompatibleScreenManifests UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:extractDeepLinksDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:processDebugMainManifest UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:processDebugManifest UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:processDebugManifestForPackage UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:processDebugResources UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:compileDebugJavaWithJavac UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugShaders UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:compileDebugShaders NO-SOURCE
[2024-10-09 17:06:44] [build-stdout] > Task :app:generateDebugAssets UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugAssets UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:compressDebugAssets UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:processDebugJavaRes NO-SOURCE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugJavaResource UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:checkDebugDuplicateClasses UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:desugarDebugFileDependencies UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeExtDexDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeLibDexDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:dexBuilderDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeProjectDexDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugJniLibFolders UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:mergeDebugNativeLibs NO-SOURCE
[2024-10-09 17:06:44] [build-stdout] > Task :app:stripDebugDebugSymbols NO-SOURCE
[2024-10-09 17:06:44] [build-stdout] > Task :app:validateSigningDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:writeDebugAppMetadata UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:writeDebugSigningConfigVersions UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:packageDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:createDebugApkListingFileRedirect UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] > Task :app:assembleDebug UP-TO-DATE
[2024-10-09 17:06:44] [build-stdout] BUILD SUCCESSFUL in 381ms
[2024-10-09 17:06:44] [build-stdout] 31 actionable tasks: 31 up-to-date
Finalizing database at /Users/owenrwang/testqlwithjava/database.
CodeQL detected code written in Java/Kotlin but could not process any of it. For more information, review our troubleshooting guide at https://gh.io/troubleshooting-code-scanning/no-source-code-seen-during-build.

The log file is as follows (it seems similar to the previous Kotlin project).
[build-tracer.log](https://github.com/user-attachments/files/17305657/build-tracer.log)
[database-create-20241009.170639.746.log](https://github.com/user-attachments/files/17305660/database-create-20241009.170639.746.log)

If I use the `--build-mode none` command on an already compiled Java project, it can successfully generate src.zip, and I can see that my code is correctly recognized in the output file of the `codeql database analyze` command. 

So, in summary, it seems there are two issues:
1.Java or Kotlin Android projects cannot recognize the source code in autobuild build mode.
2.Using `--build-mode none` can correctly recognize compiled Java source code, but compiled Kotlin source code cannot be recognized.

**Comment by Goudan932** [comment]
@igfoo Thank you very much, the problem has been resolved. It may be due to the fact that my code has already been compiled and cached, causing CodeQL to not recognize my code
